Issue #266 - Draw border around non transparent part of PNG image

kmanaseryan opened the pull request:
### Related issue:
https://github.com/viliusle/miniPaint/issues/266

### Result

Previous implementation
<img width="381" alt="Screen Shot 2021-09-25 at 14 02 16" src="https://user-images.githubusercontent.com/3371013/134767498-11a93abc-f768-4903-8e3c-6d73bcd091e3.png">

Current implementation
<img width="363" alt="Screen Shot 2021-09-25 at 14 01 22" src="https://user-images.githubusercontent.com/3371013/134767502-a419f652-716c-4570-8980-df68b3e4d1b7.png">

kmanaseryan committed:
Issue #266 - Apply styling

kmanaseryan committed:
Issue #266 - Improve border effect so that it will draw around nontransparent part of the image

kmanaseryan commented on src/js/modules/effects/borders.js:
```diff
@@ -1,107 +1,124 @@
+
+    render_post(ctx, data, layer) {
+        const size = Math.max(0, data.params.size);
+
+        let x = layer.x;
+        let y = layer.y;
+        let width = parseInt(layer.width);
+        let height = parseInt(layer.height);
+
+        //legacy check
+        if (x == null) x = 0;
+        if (y == null) y = 0;
+        if (!width) width = config.WIDTH;
+        if (!height) height = config.HEIGHT;
+
+        ctx.save();
+
+        var dArr = [-1, -1, 0, -1, 1, -1, -1, 0, 1, 0, -1, 1, 0, 1, 1, 1]; // offset array
+
+        // Draw the same image, but scaled by the border size
+        for (let i = 0; i < dArr.length; i += 2)
+            ctx.drawImage(
+                layer.link,
+                x + dArr[i] * size,
+                y + dArr[i + 1] * size
+            );
+
+        // Set the color
+        ctx.fillStyle = data.params.color;
+        // Now we will intersect the above drawn image with the rectangle below
+        // As a result we will have an object having the same shape of the original image, but scaled by the border size
+        ctx.globalCompositeOperation = "source-in";
+        ctx.fillRect(0, 0, width + size, height + size);
+
+        // Draw the original image in normal size on top of the newly created object
+        ctx.globalCompositeOperation = "source-over";
+        ctx.drawImage(layer.link, x, y);
+
+        ctx.restore();
+    }
```
@JordanMagnuson the main change has been done in this function, the rest are just styling changes.

JordanMagnuson replied:
This looks good, @kmanaseryan! I will test it out on Monday.

kmanaseryan committed:
Issue #266 - Improvements in border effect

kmanaseryan committed:
Issue #266 - Implement the border effect logic in an isolated canvas

JordanMagnuson commented:
Nice work @kmanaseryan! 

I need to do more testing of the effect, but one thing I noticed is that it seems to create fatter diagonal borders than vertical/horizontal. See https://i.imgur.com/HEpbSHf.png

(Notice also that the border around the small interior circle creates a diamond shaped-hole, rather than a circular hole). 

For comparison, here is the same image with a stroke effect applied in Photoshop:
https://i.imgur.com/Zkkh6PF.png

Here is the sample image: https://i.imgur.com/fKnV3iD.png

Anyway, nice work again. I think these are probably minor details. (Also for now don't worry about making the edges "rounded" as Photoshop does.)

kmanaseryan commented:
Thanks for the detailed feedback @JordanMagnuson. 

I think I know what are the issues, I'll take a better look into that and let you know how we can improve them.

kmanaseryan committed:
Merge branch 'master' into feature/border-effect-266

kmanaseryan committed:
Issue #266 - Revert back to prev border effect implementation

kmanaseryan committed:
Merge branch 'master' into feature/border-effect-266

kmanaseryan committed:
Issue #266 - Draw border by analyising the pixel data

kmanaseryan committed:
Issue #266 - Draw border by drawing lines through the points

- Constract all those points which are alinged on the edges of the image
- Constract separate shapes by using the points above
- Draw each shape through Canvas line drawing APIs
